dissect/target/helpers/protobuf.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, BinaryIO
+
+from dissect.cstruct.types.bytesinteger import BytesInteger
+
+if TYPE_CHECKING:
+    from dissect.cstruct import cstruct
+
+
+class ProtobufVarint(BytesInteger):
+    """Implements a protobuf integer type for dissect.cstruct that can span a variable amount of bytes.
+
+    Mainly follows the cstruct BytesInteger implementation with minor tweaks
+    to support protobuf's msb varint implementation.
+
+    Resources:
+        - https://protobuf.dev/programming-guides/encoding/
+        - https://github.com/protocolbuffers/protobuf/blob/main/python/google/protobuf/internal/decoder.py
+    """
+
+    def __init__(self, cstruct: cstruct, name: str, size: int, signed: bool, alignment: int = None):
+        self.signed = signed
+        super().__init__(cstruct, name, size, alignment)
+
+    def _read(self, stream: BinaryIO, context: dict[str, Any] = None) -> int:
+        self.size, data = read_varint(stream)
+        return decode_varint(data)
```

Comment:
```suggestion
        return read_varint(stream)
```
We don't want to change `self.size` since that's a class variable for the type. So I'd suggest merging `read_varint` and `decode_varint` into one function since we don't really end up using the size.

Reply:
I think we do need to store the size somewhere, as it seems to be required for `_read_array` to work. When I set `ProtobufVarint(size=0)` in the docker helper, we get a `MemoryError` as `_read_array` ends up in an infinite loop.

Reply:
I expect the `_read_array` to be buggy now then. The `.size` attribute is usually a static size for a single element of this type. You can see this will result in weird behaviour, since `BytesInteger` will try to do an optimized array read: https://github.com/fox-it/dissect.cstruct/blob/main/dissect/cstruct/types/bytesinteger.py#L83. I'm not sure if the current unit test catches this edge case (different sized variable integers?).

The right approach would likely be to either implement a custom `_read_array`, or to fall back on the `_read_array` of the base class: https://github.com/fox-it/dissect.cstruct/blob/main/dissect/cstruct/types/base.py#L95

Note that this will also need to be changed once https://github.com/fox-it/dissect.cstruct/pull/35 is merged, but let's cross that bridge once we get there. At least there are some unit tests :smile:

Reply:
When setting `ProtobufVarint(size=1)` in `test_protobuf.py` (just like the docker plugin does), it seems that the position of the stream is incorrectly calculated. 

```
        aaa = b"a" * 123456
        foo = cs.foo(BytesIO(b"\x00\x00\x00\x01\xc0\xc4\x07" + aaa + b"\x01\x02\x03"))
        assert foo.foo == 1
        assert foo.size == 123456
>       assert foo.bar == aaa
E       AssertionError: assert b'\xc4\x07aaa...aaaaaaaaaaaaa' == b'aaaaaaaaaaa...aaaaaaaaaaaaa'
```

As you can see the last two bytes of the `varint` are prepended to the value of `char bar[size]`. I think this is because the stream is not read 3 bytes further before parsing the next cstruct type, but just one byte. How can we tell the parent cstruct parser to seek two bytes further? 

Setting `ProtobufVarint(size=0)` in the `docker.py` helper also seems to break, but for a different reason, as the `s_len` or `m_len` is incorrectly calculated as a huge number.

Reply:
I've pushed a fix and support for reading/writing arrays for varint, including some tests.

Reply:
Thanks for the commit, this seems to fix both problems.